pkg/controller/deployment/deployment_controller.go

```diff
@@ -446,23 +475,82 @@ func (dc *DeploymentController) handleErr(err error, key interface{}) {
 	dc.queue.Forget(key)
 }
 
-// claimReplicaSets uses NewReplicaSetControllerRefManager to classify ReplicaSets
-// and adopts them if their labels match the Deployment but are missing the reference.
-// It also removes the controllerRef for ReplicaSets, whose labels no longer matches
-// the deployment.
-func (dc *DeploymentController) claimReplicaSets(deployment *extensions.Deployment) error {
-	rsList, err := dc.rsLister.ReplicaSets(deployment.Namespace).List(labels.Everything())
+// getReplicaSetsForDeployment uses ControllerRefManager to reconcile
+// ControllerRef by adopting and orphaning.
+// It returns the list of ReplicaSets that this Deployment should manage.
+func (dc *DeploymentController) getReplicaSetsForDeployment(d *extensions.Deployment) ([]*extensions.ReplicaSet, error) {
+	rsList, err := dc.rsLister.ReplicaSets(d.Namespace).List(labels.Everything())
 	if err != nil {
-		return err
+		return nil, err
+	}
+	deploymentSelector, err := metav1.LabelSelectorAsSelector(d.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("deployment %s/%s has invalid label selector: %v", d.Namespace, d.Name, err)
+	}
+	cm := controller.NewReplicaSetControllerRefManager(dc.rsControl, d, deploymentSelector, controllerKind)
+	return cm.ClaimReplicaSets(rsList)
+}
+
+// getPodMapForReplicaSets scans the list of all Pods and returns a map from
+// RS UID to Pods controlled by that RS, based on the Pod's ControllerRef.
+func (dc *DeploymentController) getPodMapForReplicaSets(namespace string, rsList []*extensions.ReplicaSet) (map[types.UID]*v1.PodList, error) {
+	// List all Pods.
+	pods, err := dc.podLister.Pods(namespace).List(labels.Everything())
```

Comment:
The pods that match in the controller ref but don't match in labels are invalid and the rs controller will eventually remove the controller ref, right? The deployment controller shouldn't care about that and it should handle only valid pods.